### PR TITLE
Test that the compiler/library builds with validate-mir

### LINKED ISSUE #105706: Test that the compiler/library builds with -Zmir-opt-level=3 -Zvalidate-mir
Because currently it doesn't. `-Zmir-opt-level=3` has a handful of problems which are reported elsewhere, but discovering problems with MIR optimizations is often as simple as building the compiler with the optimization enabled and `-Zvalidate-mir`. So we should do that in CI.

Note that all MIR opt levels are supposed to be sound, and known unsoundness is supposed to be gated behind `-Zunsound-mir-opts`, which I am not suggesting we test with.

---
Currently attempting to build the standard library with these flags results in pages of this error:
```
error: internal compiler error: broken MIR in Item(WithOptConstParam { did: DefId(0:20363 ~ core[b221]::core_simd::masks::mask_impl::{impl#6}::to_bitmask_integer), const_param_did: None }) (after pass DestinationPropagation) at bb2[2]:
                                encountered overlapping memory in `Call` terminator: _0 = <U as ReverseBits>::reverse_bits(move _0, const LANES) -> [return: bb3, unwind: bb6] 
   --> /home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/../../portable-simd/crates/core_simd/src/masks/full_masks.rs:217:13
    |    
217 |             bitmask.reverse_bits(LANES)
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |    
    = note: delayed at compiler/rustc_const_eval/src/transform/validate.rs:779:26
```
Which looks to me like https://github.com/rust-lang/rust/issues/105428

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #105706

### CODE REVIEW

**Comment by saethlin** [comment]
@jyn514 I think this is good to go now, the fix just merged

**Comment by chenyukang** [comment]
@saethlin 
Seems some UI are still failing, I rebased with master.

**Comment by saethlin** [comment]
Ah, those are not going to pass as a result of any other PR, you need to do something here to get their output back to what they're supposed to be. Probably we want to default all test suites to MIR opt level 3, _except_ for the mir-opt suite.

You'll probably also run into breakage in codegen or debuginfo tests from changing their MIR opt level, that might resolve itself when https://github.com/rust-lang/rust/pull/105577 lands.

**Comment by jyn514** [comment]
> Ah, those are not going to pass as a result of any other PR, you need to do something here to get their output back to what they're supposed to be. Probably we want to default all test suites to MIR opt level 3, except for the mir-opt suite.

That implies we need to *not* set `mir-opt-level` for the mir-opt test suite, right? So this change probably needs to happen at least partly in compiletest, not just bootstrap.

**Comment by chenyukang** [comment]
> > Ah, those are not going to pass as a result of any other PR, you need to do something here to get their output back to what they're supposed to be. Probably we want to default all test suites to MIR opt level 3, except for the mir-opt suite.
> 
> That implies we need to _not_ set `mir-opt-level` for the mir-opt test suite, right? So this change probably needs to happen at least partly in compiletest, not just bootstrap.

Is it this line make compiletest run with different options?

https://github.com/rust-lang/rust/pull/105736/files#diff-79e3e43fa59f1f75c3e925ab874d736286eb2ac7e1e08f3b7f05bae1a67f082fR67

when in `mir-opt` mode, we set `mir-opt-level` here:
https://github.com/rust-lang/rust/blob/e3961864075eaa9e855e5eec6b4f148029684539/src/tools/compiletest/src/runtest.rs#LL1952

**Comment by chenyukang** [comment]
@saethlin 
do you know why we add `mir-opt-level` for stage1 std will make fail?
```console
x test --stage 1 src/test/mir-opt
```

If so, we can not enable it for `stage1 std`.
I plan to enable it for test cases(except for those mir-opt testcases).

**Comment by saethlin** [comment]
Ah yeah I should have expected this. Let me do some experimentation locally, I'll get back to you within a day.

**Comment by chenyukang** [comment]
> Ah yeah I should have expected this. Let me do some experimentation (based on your branch) locally, I'll get back to you within a day.

Thanks, use this to reproduce it locally:
```console
./configure --set rust.validate-mir-opts=3
x test --stage 1 src/test/mir-opt
```

**Comment by saethlin** [comment]
I think we're in a sticky situation here. What I originally envisioned was a fully separate step in CI where we _only_ build and do not run any tests, we would just rely on the MIR validation. If that's possible to do, I think it is the best way to dodge the situation I will ramble on about now.

The problem you're running into is we compile the standard library ahead of time with a separate set of flags than are specified in mir-opt and codegen tests (and I suppose various other UI tests as well). So changing the build settings for the standard library will break tests when standard library functions get inlined into the tests. One other way around this is to always build the distributed standard library at the higher MIR opt level, then bless the tests. I'm told cranelift uses a higher MIR opt level for the standard library already, though based on _gestures broadly_ I don't think I would advise doing that at this moment. (This also drags in the question of what happens to users of build-std, do they get a standard library with less optimizations because the default release build is mir opt level 2? That seems bad.)

**Comment by jyn514** [comment]
I think for now we should only land this for the standard library itself; I think that would have avoided most of the issues @saethlin mentioned in the past. We can worry about doing it for the test suite and compiler later.

**Comment by jyn514** [comment]
Oh I see, even doing it for *just* the standard library causes issues. That's ... very unfortunate.

We don't run any tests on mingw-check - maybe we can add this new check to only that builder? Then the different MIR output shouldn't cause any tests to fail.

> 

> I'm told cranelift uses a higher MIR opt level for the standard library already, though based on _gestures broadly_ I don't think I would advise doing that at this moment. (This also drags in the question of what happens to users of build-std, do they get a standard library with less optimizations because the default release build is mir opt level 2? That seems bad.)

I don't think we should penalize all users just because some users wouldn't benefit. As long as we're not actively hurting build-std people, I don't think that should be a reason to avoid adding more optimizations for dist builds (Mark-Simulacrum has already said a few times people should use the official builds if they want to match performance characteristics anyway). But if you're not confident all the opt-level=2 opts are sound, we can avoid turning them on for now.

**Comment by saethlin** [comment]
`mir-opt-level=2` is the default in release, we're using it already.

`mir-opt-level=3` ICEs on a code path that doesn't _currently_ exist in the standard library but there's no reason an innocent PR could add it and become blocked: https://github.com/rust-lang/rust/issues/106141

And once that is closed, I'll do some crater kind of research. If that looks good, based on your comment here, I will propose building the standard library with more MIR opts.

**Comment by jyn514** [comment]
Looks like https://github.com/rust-lang/rust/issues/106141 was fixed :)

**Comment by saethlin** [comment]
Yup, so I have a few concerns left.

If we're going to change any defaults, we should have a crater run with `-Zmir-opt-level=3` plus `-Zvalidate-mir`.

MIR optimizations sometimes deoptimize code. This problem is one reason optimizations don't get promoted to `-Zmir-opt-level=2`. I'm not sure anyone knows which optimizations de-optimized final codegen. And these optimizations aren't confined to `-Zmir-opt-level=4`. In fact as far as I'm aware, there is no strict distinction for what goes in opt level 3 vs 4. For sure I know that the opt-level 4 pass `MultipleReturnTerminators` breaks a lot of LLVM optimizations. `DestinationPropagation` used to break optimizations of a particular `match` pattern, but that was fixed in LLVM. MIR inlining breaks some optimizations related to `mem::swap`. But also every so often, someone will complain about a missing LLVM optimization then turn up the MIR opt level and the codegen out of LLVM gets improved.

If we are going to build the standard library with a different MIR opt level, perhaps we should only do that on nightly? At least then, users could opt out (heh) if they saw a codegen issue. But then we'd have a permanent codegen difference between stable and nightly. 🤷 

Btw I'm not even on wg-mir-opt, the members probably have more thoughts about this situation than I do.

**Comment by jyn514** [comment]
Oh right, you suggested building unconditionally with opt-level=2, I'd forgotten.

I think for now it might be enough to build std with opt-level=3 on mingw-check, without running tests, and worry about changing the default in a follow-up PR. I'm not super enthusiastic about needing to fix every possible size/perf regression before doing any validation that validate-mir actually works.

**Comment by chenyukang** [comment]
> What I originally envisioned was a fully separate step in CI where we _only_ build and do not run any tests, we would just rely on the MIR validation. If that's possible to do, I think it is the best way to dodge the situation I will ramble on about now.

@jyn514 
Is this doable?
If we only build std with the MIR validation purpose, then rebuild it with a different opt level before testing.
Of course, this will take more time to finish CI.

**Comment by jyn514** [comment]
@chenyukang mingw-check *only* checks and doesn't build the standard library; it also finishes a lot sooner than other jobs. I think making it a little longer is ok :) std doesn't take long to buildm

**Comment by jyn514** [comment]
@chenyukang are you planning to keep working on this? is it clear what the next steps are?

**Comment by chenyukang** [comment]
> @chenyukang are you planning to keep working on this? is it clear what the next steps are?

yes, I will pick it up later.

File: src/bootstrap/builder.rs [hunk 1]
```diff
@@ -1915,6 +1915,13 @@ impl<'a> Builder<'a> {
             }
         }
 
+        if matches!(mode, Mode::Std) && compiler.stage == 0 {
```
Comment (jyn514) [inline src/bootstrap/builder.rs:1918]: Why this `stage == 0` check?

File: src/ci/docker/host-x86_64/mingw-check/validate-mir.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+echo $'[rust]\nvalidate-mir-opts = 3' > /tmp/validate-mir.toml
+if ! python3 ../x.py build --stage 0 library/std --config /tmp/validate-mir.toml; then
+    echo "failed validate mir!"
+    exit 1
+fi
+rm /tmp/validate-mir.toml
+echo "finished validate mir"
```
Comment (jyn514) [inline src/ci/docker/host-x86_64/mingw-check/validate-mir.sh:9]: Hmm, I'm not sure why this script and the change to `SCRIPT` were necessary? bootstrap is already doing the logic for you, you don't need to make sure to invoke it only on stage 0.

**Review by jyn514 — CHANGES_REQUESTED** [review]

File: src/ci/docker/host-x86_64/mingw-check/validate-mir.sh (hunk 2 above)
Comment (chenyukang) [reply src/ci/docker/host-x86_64/mingw-check/validate-mir.sh:9]: you means add `RUST_CONFIGURE_ARGS` at here?
https://github.com/rust-lang/rust/blob/07c993eba8b76eae497e98433ae075b00f01be10/.github/workflows/ci.yml#L52

Comment (jyn514) [reply src/ci/docker/host-x86_64/mingw-check/validate-mir.sh:9]: No, in the dockerfile itself.

Comment (chenyukang) [reply src/ci/docker/host-x86_64/mingw-check/validate-mir.sh]: updated!

File: src/ci/docker/host-x86_64/mingw-check/Dockerfile [hunk 3]
```diff
@@ -39,7 +41,7 @@ COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
 ENV SCRIPT python3 ../x.py --stage 2 test src/tools/expand-yaml-anchors && \
            python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu --all-targets && \
-           python3 ../x.py build --stage 0 src/tools/build-manifest && \
+           python3 ../x.py build --stage 0 src/tools/build-manifest library/std && \
```
Comment (chenyukang) [inline src/ci/docker/host-x86_64/mingw-check/Dockerfile:44]: Not sure whether we need to invoke building `library/std` explicitly.

Comment (jyn514) [reply src/ci/docker/host-x86_64/mingw-check/Dockerfile:44]: You don't have to :) it's already built automatically here: https://github.com/rust-lang/rust/blob/7d3457bb1220df1afac72eab41806c82cb67d250/src/ci/docker/host-x86_64/mingw-check/Dockerfile#L46

**Comment by jyn514** [comment]
This looks good to me! can you just verify by looking at the CI logs that `validate-mir-opts` is actually being passed to configure? github actions is being annoying and won't let me look at the full logs until the run finishes 😞 but it will look something like `configure: build.configure-args := ['--set', 'rust.validate-mir-opts']`

**Comment by jyn514** [comment]
ah, here it is: https://github.com/rust-lang/rust/actions/runs/4266886160/jobs/7427913259#step:26:1091

@bors r+

**Comment by jyn514** [comment]
oh sorry, can you fix https://github.com/rust-lang/rust/pull/105736#discussion_r1117779315 first?

@bors r- delegate=chenyukang

(please use `@bors r=jyn514` when you've fixed it :)

**Pull request merged**